src/main.c

```diff
@@ -115,9 +118,12 @@ static bool check_dfu_mode(void)
 int main(void)
 {
   board_init();
-
   TU_LOG1("TinyUF2\r\n");
 
+  // Flash needs to be initialized to check for a valid image
+  TU_LOG2("Flash init...\r\n");
+  board_flash_init();

```

Comment:
[MANDATORY] you cannot change the order of function in main.c , this will affect other ports. If it is needed, added init function that must be called in either board_init() or board_dfu_init()

Reply:
This is critical for this port, and may be needed in other ports as well.  You cannot check the validity of an app in flash without initializing the flash interface.  This order is better for the iMXRT10xx port as well.  It is less critical for most devices with internal flash, except for LPC55 which has bus error when reading erased flash.

Reply:
if `FLASH_Init` is critical within lpc55 that must be called, you could just place it in the `board_init()` for now. `board_flash_init()` is mostly for preparing flash for writing op, which is why I currently put it after `board_dfu_init()`

Reply:
Does it cause any harm to call flash init first?
If you are running from RAM, you would need to initialize the flash before jumping to the application.  This is a difference between pushing code into RAM with sdphost vs letting the boot ROM copy the code into RAM on i.MXRT.  The boot ROM pre configures the flash for reading, but you need to do this manually if you load with sdphost.   If they load the image with sdphost, they will most often be updating the bootloader, but it seems safer to initialize the flash so the behavior is the same and we do not prevent them from jumping to an image in flash.  

I can make this change if you think there will be adverse side effects for other platforms.

Reply:
What if instead of calling board_flash_init in main, we call it from board_init, or board_dfu_init depending on the platform requirements?  
This would require more changes, but would be cleaner than implementing with #ifdef

Reply:
@nxf58843 that is possible as well but not in the scope of this PR. If you are willing to make the changes and test it on all other platforms, feel free to make a separate PR for the changes.

Reply:
I moved boar_flash_init into board init for the mimxrt10xx and lpc55 platforms, and added `#ifndef EARLY_FLASH_INIT` for other platforms in main.c to get this PR through.  This can be simplified in a future PR.

Reply:
thank you for the changes, though I don't like the introduction of `EARLY_FLASH_INIT`, we should completely remove the `board_flash_init()` making it the internal port function that either is called within board_init() or board_dfu_init() if needed.

Reply:
PS: I probably didn't make myself clear enough, sorry for that. I mean we should completely remove the board_flash_init() by making it part of either board_init() (early) or board_dfu_init() (late) decided by each port. You don't have to it as part of this PR. just get it passed while keeping the current order of execution by having extern internal_flash_init() to call `FLASH_Init()` and implement `board_flash_init()` as empty. I will do a cross-port PR to refactor it as follow-up later on.

Reply:
Ok, technically speaking, initializing flash earlier for i.MXRT10xx would make the firmware behave more consistently if it gets copied to RAM in internal boot mode by the ROM, or if it is pushed in through serial downloader boot mode, since the serial downloader boot mode would not initialize the FlexSPI interface.  But, in this case, behaving differently is a happy accident.  Not initializing the flash before checking the app validity means that it will never find a valid app when you push the image in through serial downloader mode, but that is probably the safest way to ensure that you can recover the bootloader image.  If you are pushing an image in through serial downloader mode, the odds are that you want to update the bootloader, and double tap reset will not work because it will go back to serial downloader mode.  Thus, to be able to update when you have a valid app and a broken bootloader, the safest thing to do is to always enter dfu mode whenever the image was loaded through the serial downloader, and initializing the flash after the validity check accomplishes that.  I'll work on cleaning this up.

Reply:
wow, this is indeed a happy accident, look like we are very lucky for this one :)